fast_llm/tools/prepare_dataset.py

```diff
@@ -0,0 +1,359 @@
+        config.validate()
+        self._config = config
+
+    def run(self) -> None:
+        raise NotImplementedError
+
+
+@config_class
+class GPTDatasetConfig(Config):
+    name_or_path: str = Field(
+        desc="Name or path of the dataset.",
+        hint=FieldHint.core,
+    )
+    config_name: None | str = Field(
+        default=None,
+        desc="Specific configuration name for the dataset.",
+        hint=FieldHint.optional,
+    )
+    split: str = Field(
+        default="train",
+        desc="Split of the dataset to use.",
+        hint=FieldHint.optional,
+    )
+    field: str = Field(
+        default="text",
+        desc="Field of the dataset to use.",
+        hint=FieldHint.optional,
+    )
+    data_type: DataType = Field(
+        default=None,
+        desc="Data type of the dataset field.",
+        hint=FieldHint.derived,
+    )
+    trust_remote_code: bool = Field(
+        default=False,
+        desc="Trust remote code when downloading the dataset.",
+        hint=FieldHint.optional,
+    )
+    disable_disk_space_check: bool = Field(
+        default=False,
+        desc="Disable disk space check. Useful for environments where disk space is not accurately reported.",
+        hint=FieldHint.optional,
+    )
+
+
+@config_class()
+class GPTDatasetPreparatorConfig(DatasetPreparatorConfig):
+    _abstract = False
+    model_name: typing.ClassVar[str] = "gpt"
+
+    tokens_per_shard: int = Field(
+        default=1_000_000_000,
+        desc="Approximate number of tokens per shard.",
+        hint=FieldHint.feature,
+        valid=check_field(Assert.geq, 100_000),
+    )
+    loading_workers: int = Field(
+        default=1,
+        desc="Number of workers in load_dataset() call.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 1),
+    )
+    tokenize_workers: int = Field(
+        default=1,
+        desc="Number of workers for tokenization.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 1),
+    )
+    saving_workers: int = Field(
+        default=1,
+        desc="Number of processes for saving the data.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 1),
+    )
+    remove_downloads: bool = Field(
+        default=False,
+        desc="Remove downloaded dataset after processing.",
+        hint=FieldHint.optional,
+    )
+    dataset: GPTDatasetConfig = Field(
+        default_factory=GPTDatasetConfig,
+        desc="Configuration for the dataset.",
+        hint=FieldHint.feature,
+    )
+    tokenizer: TokenizerConfig = Field(
+        default_factory=TokenizerConfig,
+        desc="Configuration for the tokenizer.",
+        hint=FieldHint.feature,
+    )
+    _tokenizer: Tokenizer = Field(
+        init=False,
+        desc="The tokenizer instance.",
+        hint=FieldHint.derived,
+    )
+
+    def _validate(self):
+        Assert.not_none(self.tokenizer.path)
+        self._tokenizer = Tokenizer(config=self.tokenizer)
```

Comment:
This is doing some work and reading files, so it shouldn't go in `_validate` (not runnable on remote and requires dependencies). The tokenizer can be built in the preparator.

Reply:
Hm, not sure how I can both, derive the data type based on the tokenizer length and not import the tokenizer 😉
I can make it such that if the tokenizer dependency cannot be fulfilled the validation fails unless the data type is user specified in the config.

Reply:
Can we do validation without knowing them? I think we could have `data_type: DataType | None = None` and derive the actual value in `run`.